gapis/api/gles/api/programs_and_shaders.api

```diff
@@ -2380,9 +2380,11 @@ cmd void glUseProgram(ProgramId program) {
   } else {
     p := GetProgramOrError(program)
     if p.LinkStatus == GL_FALSE { glErrorInvalidOperation() }
-    AdjustProgramUseCount(p, +1)
-    AdjustProgramUseCount(ctx.Bound.Program, -1)
-    ctx.Bound.Program = p
+    if p != ctx.Bound.Program {
```

Comment:
Doesn't this CREATE a dependency between glUseProgram calls?
Now we are reading ctx.Bound.Program, rather than just writing it.

Reply:
Good point! I think this changes is still valid, but my explanation was not clear.

This change makes sense because a draw call will only depend on the latest UseProgram that wrote to ctx.Bound.Program. Consider:
```
1. glUseProgram(42)
2. glUseProgram(42)
3. glDraw()
```

Without the change, glUseProgram() calls AdjustProgramUseCount which performs a read-write to program.UseCount. This creates a dependency between 1 and 2. The draw call in 3 depends on 2, because 2 is the latest who wrote to ctx.Bound.Program.

With the change, there is still a dependency between 1 and 2, as you pointed out. Yet, 2 does NOT write on ctx.Bound.Program. So the draw call 3 depends on **1** this time, and not on 2. Thanks to DCE, 2 will be removed.

Does that make sense?

I will make sure the commit message clarifies this.

Reply:
I think, though, the problem is that if:
```
1. glUseProgram(1)
2. glUseProgram(2)
3. glDraw()
```
It will now no longer eliminate 1.

Can we make the dependency tracking smarter and have it ignore writes of the same value? i.e. outside of the API files?

Reply:
Ah yes indeed, thanks for the clarification.

See #2952 for a discussion about whether the dependency graph ignoring writes that do not modify the state.

Reply:
Is the count something that should be tracked for dependencies, or is the count only something that is used internally? If the latter, we should not track  it.

Reply:
The count is relevant to know whether the program can be deleted:
```
sub void AdjustProgramUseCount(ref!Program p, s32 delta) {
  ctx := GetContext()
  if p != null {
    p.UseCount += delta
    assert(p.UseCount >= 0)
    // Delete if it was marked for deletion and it is no longer used anywhere
    if p.DeleteStatus && (p.UseCount == 0) {
      for _ , _ , s in p.Shaders {
        AdjustShaderAttachCount(s, -1)
      }
      delete(ctx.Objects.Programs, p.ID)
    }
  }
}
```